lib/queues.js

```diff
@@ -611,5 +608,58 @@ module.exports = class Queues {
     // Lazy load configuration.
     return this._ironium.configPromise;
   }
-};
 
+  // Resolves to the maximum amount of jobs a queue should try to reserve.
+  // Promise doesn't resolve until capacity is available.
+  getReserveCapacity() {
+    return this._calculateReserveCapacity()
+      .then(capacity => {
+        if (capacity > 0)
+          return capacity;
+        else {
+          return new Promise(resolve => {
+            this._waitingForReserve.push(resolve);
+          });
+        }
+      });
+  }
+
+  // Tries to allocate requested capacity to run jobs.
+  // Resolves to the actual capacity allocated.
+  // The difference between requested and allocated is
+  // the amount of jobs that should be returned to the queue.
+  allocateCapacity(requestedCapacity) {
+    return this.configPromise
+      .then(config => {
+        const totalCapacity      = config.concurrency - this._activeJobs;
+        const allocatedCapacity  = Math.min(totalCapacity, requestedCapacity);
+        this._activeJobs        += allocatedCapacity;
+        return allocatedCapacity;
+      });
+  }
+
+  // Frees capacity to run jobs.
+  freeCapacity(count) {
```

Comment:
For emptying the array (actually, just realized that passing 0 is not needed). When the queue is blocked again, a new resolve function will be pushed to the array.